lib/alchemy/upgrader/four_point_zero.rb

```diff
@@ -1,3 +1,118 @@
+        build_new_elements(old_element).inject(ne) do |ne, element_from_contents|
+          ne << element_from_contents
+        end
+      end
+      new_elements = new_elements.uniq.sort {|a, b| a['name'] <=> b['name'] }
+      puts 'done.'
+
+      print 'Writing new `config/alchemy/elements.yml` ... '
+      write_config(new_elements)
+      puts 'done.'
+
+      print 'Removing `render_new_content_link` helper from editor partials ... '
+      remove_new_content_link_from_editor_partials
+      puts 'done.'
+
+      puts 'Adding hints for rendering nested elements to your views: '
+      add_render_nested_elements_hints
+    end
+  end
+
+  private
+
+  def backup_config
+    print "Copying existing config file to config/alchemy/elements.yml.old ... "
+    FileUtils.copy  Rails.root.join('config', 'alchemy', 'elements.yml'),
+                    Rails.root.join('config', 'alchemy', 'elements.yml.old')
+    puts "done."
+  end
+
+  def read_config
+    print "Reading config/alchemy/elements.yml ... "
+    old_config_file = Rails.root.join('config', 'alchemy', 'elements.yml')
+    config = YAML.load_file(old_config_file)
+    puts "done."
+    config
+  end
+
+  def write_config(config)
+    backup_config
+    File.open(Rails.root.join('config', 'alchemy', 'elements.yml'), "w") do |f|
+      f.write config.to_yaml
+    end
+  end
+
+  def modify_old_element(element)
+    nestable_elements = element['available_contents'].map do |content|
+      "addable_#{content['name']}"
+    end
+    element.delete('available_contents') # Hashes are mutable. Welcome!
+    element['nestable_elements'] = nestable_elements
+    element
+  end
+
+  def remove_new_content_link_from_editor_partials
+    editor_partials = Dir.glob(Rails.root.join('app', 'views', 'alchemy', 'elements', '*_editor*'))
+    system "sed -i '' '/^.*render_new_content_link.*$/d' #{editor_partials.join(' ')}"
+    system "sed -i '' '/^.*label_and_remove_link.*$/d' #{editor_partials.join(' ')}"
+  end
+
+  def build_new_elements(element)
+    element['available_contents'].inject([]) do |collection, content|
+      collection << build_new_element(content)
+    end
+  end
+
+  def build_new_element(content)
+    # All nestable elements are deletable
+    content['settings'].delete('deletable') if content['settings']
+    # If the only content is not present, the whole things doesn't make much sense
+    content['validate'] = ['presence']
+    content.delete('settings') if content['settings'] == {}
+    {
+      'name' => "addable_#{content['name']}",
+      'contents' => [content]
+    }
+  end
+
+  def add_render_nested_elements_hints
+    erb_views = Dir.glob(Rails.root.join('app', 'views', 'alchemy', 'elements', '*_view.html.erb'))
+    haml_slim_views = Dir.glob(Rails.root.join('app', 'views', 'alchemy', 'elements', '*_view.html.haml')) +
+                      Dir.glob(Rails.root.join('app', 'views', 'alchemy', 'elements', '*_view.html.slim'))
```

Comment:
Metrics/LineLength: Line is too long. [106/100]<br>Style/MultilineOperationIndentation: Use 2 (not 18) spaces for indenting an expression in an assignment spanning multiple lines.